### DISCUSSION TITLE
what will be the closest replacement for @auto_move_data?

### DISCUSSION THREAD

**Discussion by immanuelweber** [Lightning-AI/pytorch-lightning#7588] [Lightning Trainer API: Trainer, LightningModule, LightningDataModule] [answered]
I really liked @auto_move_data as it gave me a nice and easy way to pass (a single) input to a model, without caring about moving data structures to devices, but now it will be removed 😢 
I realize that you want people to use trainer.predict() but this requires a dataloader, right?, which at least for me oftentimes is overly complicated while developing.
As far as I see the dataloader/collate/transforms also have to be different from the normal ones, since these typically provide batches with inputs and targets, whereas the forward function only takes inputs.
Likely I'm missing something, can you maybe give some hints about this?

**Comment by immanuelweber** [comment, marked as answer]
After writing this I realized that I can use something like this: 
```
model = model.eval()
i, _ = next(iter(data.test_dataloader()))
with th.no_grad():
    outputs = model.forward(model.transfer_batch_to_device(i))
```
But I'm still curious about the intended future ways of doing this....

**Reply by carmocca** [reply]
Hi!

As you mentioned, `@auto_move_data` was deprecated to encourage `trainer.predict` use. We noticed many users incorrectly using `@auto_move_data` or adding it to many functions thinking It is necessary.

Feel free to copy the current implementation to have the decorator, but as you noticed, it is just a wrapper for `transfer_batch_to_device`.

Here is the implementation for any future readers:

https://github.com/PyTorchLightning/pytorch-lightning/blob/cf5dc04db4a84e31d8c360c350b3896b67b38aeb/pytorch_lightning/core/decorators.py#L22-L64

**Reply by immanuelweber** [reply]
all right, thank you for the clarification.
